### ISSUE TITLE
Redis client retry logic is not correct

### ISSUE THREAD

**Issue by ycraaron** [andymccurdy/redis-py#1022]
## Steps to reproduce
https://github.com/andymccurdy/redis-py/blob/413e802f8fe390738bc1e89716083586e0269fd0/redis/client.py#L670-L675

During debugging another issue, I find that the logic here is a bit strange.
The senario:
2 different types of errors ConnectionError and TimeoutError will be caught here, and from the following if clause, isinstance(TimoutError) is added as a condition, so I am assuming that the expected behavior for retry is:
1. The user sets the retry_on_timeout to True.
2. It's a TimeoutError.

## Expected behavior
The expect behavior can be interpreted with a truth table(T means do retry, F means not):

|      TimeoutError\retry | true | false | 
|-------|------|:-----:|
| true  | T    |   F   |
| false | F    |   F   |
## Actual behavior
The actual behavior can be interpreted with a truth table(T means do retry, F means not):

|      TimeoutError\retry | true | false | 
|-------|------|:-----:|
| true  | T    |   F   |
| false | T    |   T   |

## Conclusion
If my understanding of the expected behavior is correct, then we need to fix those unwanted retry.

Here is a script for validating the current behavior and behavior after the fix:
```python
# script for validating the current and fixed behavior truth table

do_retry =          [True, True, False, False]
is_timeout_error =  [False,True, True,  False]
zipped = list(zip(do_retry, is_timeout_error))

# current behaviour:
print("----------------------------------------")
print("Current behaviour: ")
print("----------------------------------------")
for do_retry, is_timeout_error in zipped:
    print("Condition: retry-{}, timeout-{}".format(do_retry, is_timeout_error))
    if not do_retry and is_timeout_error:
        print("Result: not retry")
    else:
        print("Result: retry")

# behaviour after fix:
print("----------------------------------------")
print("Fixed behaviour: ")
print("----------------------------------------")
for do_retry, is_timeout_error in zipped:
    print("Condition: retry: {}, timeout: {}".format(do_retry, is_timeout_error))
    if not (do_retry and is_timeout_error):
        print("Result: not retry")
    else:
        print("Result: retry")
print("----------------------------------------")
```
Script output:

```
----------------------------------------
Current behaviour:
----------------------------------------
Condition: retry_on_timeout-True, timeout_error-False
Result: retry
Condition: retry_on_timeout-True, timeout_error-True
Result: retry
Condition: retry_on_timeout-False, timeout_error-True
Result: not retry
Condition: retry_on_timeout-False, timeout_error-False
Result: retry
----------------------------------------
Fixed behaviour:
----------------------------------------
Condition: retry_on_timeout: True, timeout_error: False
Result: not retry
Condition: retry_on_timeout: True, timeout_error: True
Result: retry
Condition: retry_on_timeout: False, timeout_error: True
Result: not retry
Condition: retry_on_timeout: False, timeout_error: False
Result: not retry
----------------------------------------
```

**Comment by tzickel** [comment]
Question, how popular is it for client side libraries that do modifications to automatic retry while hiding it from the end user ?
HTPT POST (and requests) forbid it, I think cassandra / pymemcache don't do it, maybe the default should have it disabled ?
https://github.com/andymccurdy/redis-py/issues/977

**Comment by ycraaron** [comment]
I am personally not a fan of this kind of hidden operations, not to mention making it a default behaviour.
As from my experience, most of the RDBMS client side libraries I use won't do this, cassandra doesn't do this. I am not quite familiar with pymemcache, but nice to hear that it doesn't have this auth-retry which sometimes could be the culprit of some super strange behaviours.

I agree that it's better to have it disabled by default. @tzickel

**Comment by weishen** [comment]
@ycraaron just a bit curous,  in line 671,  the connection has been closed, why we still do a retry?

**Comment by andymccurdy** [comment]
I believe this should be resolved. 3.1.0 fixed the faulty retry logic and 3.2.0 improved the ConnectionPool in that it will only return healthy connections now.
